tfhe/src/integer/server_key/radix_parallel/cmux.rs

```diff
@@ -0,0 +1,232 @@
+    ///
+    /// // Decrypt:
+    /// let dec: u8 = cks.decrypt(&ct_res);
+    /// assert_eq!(if a >= 66 { a } else { b }, dec);
+    /// assert_ne!(ct_a, ct_res);
+    /// assert_ne!(ct_b, ct_res);
+    /// ```
+    pub fn smart_if_then_else_parallelized(
+        &self,
+        condition: &mut RadixCiphertext,
+        true_ct: &mut RadixCiphertext,
+        false_ct: &mut RadixCiphertext,
+    ) -> RadixCiphertext {
+        let mut ct_refs = [condition, true_ct, false_ct];
+
+        ct_refs.par_iter_mut().for_each(|ct_ref| {
+            if !ct_ref.block_carries_are_empty() {
+                self.full_propagate_parallelized(ct_ref);
+            }
+        });
+
+        let [condition, true_ct, false_ct] = ct_refs;
+        self.unchecked_if_then_else_parallelized(condition, true_ct, false_ct)
+    }
+
+    pub(crate) fn unchecked_programmable_if_then_else_parallelized<F>(
+        &self,
+        condition_block: &crate::shortint::Ciphertext,
+        true_ct: &RadixCiphertext,
+        false_ct: &RadixCiphertext,
+        predicate: F,
+    ) -> RadixCiphertext
+    where
+        F: Fn(u64) -> bool + Send + Sync + Copy,
+    {
+        let inverted_predicate = |x| !predicate(x);
+
+        // Although our mul algorithm has special path for when rhs or lhs is a boolean value,
+        // we don't call it as for the ct_false we would need an extra pbs to 'invert' the
+        // ciphertext from true to false.
+        let (mut true_ct, false_ct) = rayon::join(
+            move || {
+                let mut true_ct = true_ct.clone();
+                self.zero_out_if(&mut true_ct, condition_block, inverted_predicate);
+                true_ct
+            },
+            move || {
+                let mut false_ct = false_ct.clone();
+                self.zero_out_if(&mut false_ct, condition_block, predicate);
+                false_ct
+            },
+        );
+        // If the condition was true, true_ct will have kept its value and false_ct will be 0
+        // If the condition was false, true_ct will be 0 and false_ct will have kept its value
+        true_ct
+            .blocks
+            .par_iter_mut()
+            .zip(false_ct.blocks.par_iter())
+            .for_each(|(lhs_block, rhs_block)| {
+                self.key.unchecked_add_assign(lhs_block, rhs_block);
+                self.key.message_extract_assign(lhs_block)
+            });
+        true_ct
+    }
+
+    /// This function takes a ciphertext encrypting any integer value
+    /// and block encrypting a boolean value (0 or 1).
+    ///
+    /// The input integer ciphertext will have all its block zeroed if condition_block
+    /// encrypts 0, otherwise each block keeps its value.
+    pub(crate) fn zero_out_if_condition_is_false(
+        &self,
+        ct: &mut RadixCiphertext,
+        condition_block: &crate::shortint::Ciphertext,
+    ) {
+        assert!(condition_block.degree <= Degree(1));
```

Comment:
I think here we have a case where that ord derive provides something: when the assert fails
the message is
```
integer::server_key::radix_parallel::tests::test_integer_default_mul_param_multi_bit_message_3_carry_3_group_2_ks_pbs' panicked at 'assertion failed: `(left == right)`
  left: `Degree(0)`,
 right: `Degree(1)`', tfhe/src/integer/server_key/radix_parallel/cmux.rs:195:9
```

instead of 

```
integer::server_key::radix_parallel::tests::test_integer_default_mul_param_multi_bit_message_3_carry_3_group_2_ks_pbs' panicked at 'assertion failed: `(left == right)`
  left: `0`,
 right: `1`', tfhe/src/integer/server_key/radix_parallel/cmux.rs:195:9
```

Reply:
it is well defined for all our new types I guess, I tend to add messages with the debug print anyways in core, so I guess it explains why I did not really have issues, the error message has the pretty print, I don't think it's going to help particularly as someone using the .0 to check equality would likely not import the new type to have == Degree(0) and therefore would be missing out on the debug pretty print anyways. It's a coding standard.

Ok to add the derive to leave the choice I guess 🤷‍♂️

Reply:
removed it